Feature/add search

Clement3 opened the pull request.

Clement3 committed:
Add Search

Clement3 committed:
Remove Selected Fields

- This block us to manually select fields on joined table

System-Glitch commented on validation.go:
```diff
@@ -87,6 +99,7 @@ func ApplyValidation(set validation.RuleSet) {
 	set["fields"] = validation.List{"string"}
 	set["page"] = validation.List{"integer", "min:1"}
 	set["per_page"] = validation.List{"integer", "between:1,500"}
+	set["search"] = validation.List{"string"}
```
You are not using your custom validation rule for "search".

System-Glitch commented on validation.go:
```diff
@@ -74,6 +78,14 @@ func validateJoin(ctx *validation.Context) bool {
+	}
+
+	return false
+}
```
This validation rule should convert the string to a `Search` structure.

System-Glitch commented on validation.go:
```diff
@@ -18,9 +18,13 @@ func init() {
 	lang.SetDefaultValidationRule("filter.array", "The filter format is invalid.")
 	lang.SetDefaultValidationRule("join.array", "The join format is invalid.")
 	lang.SetDefaultValidationRule("sort.array", "The sort format is invalid.")
+	lang.SetDefaultValidationRule("search.string", "The search format is invalid.")
```
```suggestion
	lang.SetDefaultValidationRule("search", "The search format is invalid.")
```
This rule is not type-dependent so the ".string" part is not necessary.

System-Glitch commented on validation.go:
```diff
@@ -102,6 +115,7 @@ func ApplyValidationRules(set *validation.Rules) {
 	set.Fields["fields"] = &validation.Field{Rules: []*validation.Rule{{Name: "string"}}}
 	set.Fields["page"] = &validation.Field{Rules: []*validation.Rule{{Name: "integer"}, {Name: "min", Params: []string{"1"}}}}
 	set.Fields["per_page"] = &validation.Field{Rules: []*validation.Rule{{Name: "integer"}, {Name: "between", Params: []string{"1", "500"}}}}
+	set.Fields["search"] = &validation.Field{Rules: []*validation.Rule{{Name: "string"}}}
```
Same here, you are not using your custom validation rule for "search".

System-Glitch commented on search.go:
```diff
@@ -0,0 +1,30 @@
+	return func(tx *gorm.DB) *gorm.DB {
+		tableWithColumn := []string{tx.Statement.Quote(modelIdentity.TableName), tx.Statement.Quote(s.Field)}
+
+		return tx.Or("(lower(CAST("+ strings.Join(tableWithColumn, ".") + " AS TEXT))) LIKE lower(?)", "%" + helper.EscapeLike(s.Query) + "%")
```
To avoid conflict with filters, I think it would be better to use a [grouped condition](https://gorm.io/docs/advanced_query.html#Group-Conditions).

That also means that search would generate only one scope instead of n scopes (n being the number of `FieldsSearch`).

System-Glitch replied:
I think the feature would be more flexible if it used an `Operator` instead of hardcoded condition (that may not work with all database engines). The `Operator` could be specified by the client through the query param.

System-Glitch commented on settings.go:
```diff
@@ -23,6 +23,11 @@ type Settings struct {
+	DisableSearch bool
+
+	// FieldsSearch allows search for these fields
+	FieldsSearch []string
```
For this feature, I would see two approaches:
- Let the client define the fields used for the search (and prevent searching on fields that are in the blacklist)
- Hardcode the fields used for the search on the server-side (what you are doing in this case)

The second option requires a bit more setup on the server-side and wouldn't work at all without providing custom settings. I think the first option offers more flexibility.

System-Glitch commented on settings.go:
```diff
@@ -109,8 +138,6 @@ func (s *Settings) Scope(db *gorm.DB, request *goyave.Request, dest interface{})
 			fields = modelIdentity.addForeignKeys(fields)
 		}
 		paginator.DB = paginator.DB.Scopes(selectScope(modelIdentity, modelIdentity.cleanColumns(fields, s.FieldsBlacklist)))
-	} else {
-		paginator.DB = paginator.DB.Scopes(selectScope(modelIdentity, s.getSelectableFields(modelIdentity.Columns)))
```
Why is it removed? This prevents blacklist from working properly when no "fields" query param is provided.

Clement3 committed:
Search - Use operator to let developer decide for the search query

- Set all fields if no SearchFields
- Fix: Use Search Validation
- Add the Statement.Selects with a bool for the Join

Clement3 committed:
Filters: Add Bool operators ( $istrue, $isfalse )

System-Glitch commented on filter.go:
```diff
@@ -110,7 +110,7 @@ func join(tx *gorm.DB, joinName string, modelIdentity *modelIdentity) *gorm.DB {
 }
 
-func selectScope(modelIdentity *modelIdentity, fields []string) func(*gorm.DB) *gorm.DB {
+func selectScope(modelIdentity *modelIdentity, fields []string, withSelects bool) func(*gorm.DB) *gorm.DB {
```
`append` or `override` instead of `withSelects` would be more expressive in my opinion.

System-Glitch commented on operator.go:
```diff
@@ -67,6 +67,18 @@ var (
 		},
+		"$istrue": {
+			Function: func(tx *gorm.DB, filter *Filter, column string) *gorm.DB {
+				return filter.Where(tx, column+" IS TRUE")
```
Check that it cannot generate a database error if `column` is not `boolean`. If it does, a safety check would be needed.

System-Glitch commented on reflect.go:
```diff
@@ -84,7 +83,6 @@ func (i *modelIdentity) addForeignKeys(fields []string) []string {
 	for _, r := range i.Relations {
 		for _, k := range r.LocalKeys {
 			if !helper.ContainsStr(fields, k) {
-				fmt.Println("add", k)
```
Nice catch. Thank you for removing it!

System-Glitch commented on search.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func (s *Search) Scopes(settings *Settings, modelIdentity *modelIdentity) func(*gorm.DB) *gorm.DB {
+	var fields []string
+
+	// Remove columns that not exist in the table
+	for _, field := range s.Fields {
+		_, ok := modelIdentity.Columns[field]
+		if ok {
+			fields = append(fields, field)
+		}
+	}
+
+	if len(fields) == 0 {
+		return nil
+	}
```
Because the fields list is defined server-side, it's the developer responsibility to not add non-existing columns, so this piece of code is not mandatory.

Clement3 replied:
We should leave this security because the fields defined by the developer in the parameters are just strings, the possibility of an error is quite large but adding an error message without blocking the execution of the request would be beneficial for the user but also for the developer who uses the project.
What do you think ?

System-Glitch replied:
A warning would be nice indeed. Use `goyave.ErrLogger` to print something like "goyave.dev/filter warning: used non-existing column in search".

System-Glitch replied:
Without this security, using a non-existing column will generate a database error so the developer will be aware of their mistake.
In a finished application, this kind of scenario should never happen so I think a crash is appropriate.

System-Glitch commented on search.go:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	return func(tx *gorm.DB) *gorm.DB {
+		searchQuery := tx.Session(&gorm.Session{})
```
Does it remove select clauses entirely from this tx? I think the [`NewDB`](https://gorm.io/docs/session.html#NewDB) option is needed here. In GORM code I can see that if `NewDB` is `false`, the `tx.Statement` is cloned, so it may cause issues.

System-Glitch commented on settings.go:
```diff
@@ -74,6 +80,28 @@ func (s *Settings) Scope(db *gorm.DB, request *goyave.Request, dest interface{})
+		query, ok := request.Data["search"].(string)
+		if ok {
+			if len(s.FieldsSearch) == 0 {
+				s.FieldsSearch = s.getSelectableFields(modelIdentity.Columns)
```
Don't re-assign to `Settings`. This structure may potentially be used concurrently. Use a temporary variable instead.

System-Glitch commented on settings.go:
```diff
@@ -74,6 +80,28 @@ func (s *Settings) Scope(db *gorm.DB, request *goyave.Request, dest interface{})
+					Query:  query,
+				}
+
+				scope := search.Scopes(s, modelIdentity)
+
+				if scope != nil {
+					db = db.Scopes(scope)
+				}
```
```suggestion
				if scope := search.Scopes(s, modelIdentity); scope != nil {
					db = db.Scopes(scope)
				}
```

System-Glitch commented on search.go:
```diff
@@ -0,0 +1,43 @@
+	Query  string
+}
+
+func (s *Search) Scopes(settings *Settings, modelIdentity *modelIdentity) func(*gorm.DB) *gorm.DB {
```
```suggestion
func (s *Search) Scope(settings *Settings, modelIdentity *modelIdentity) func(*gorm.DB) *gorm.DB {
```
This function only returns one scope so it can be called `Scope` instead of `Scopes`.

System-Glitch commented on search.go:
```diff
@@ -0,0 +1,43 @@
+		for _, field := range fields {
+			filter := &Filter{
+				Field:    field,
+				Operator: settings.SearchOperator,
```
Use operator `$cont` by default if `settings.SearchOperator` is `nil` (not defined in `Settings`).

System-Glitch commented on settings.go:
```diff
@@ -23,6 +23,12 @@ type Settings struct {
+
+	// FieldsSearch allows search for these fields
+	FieldsSearch   []string
+	SearchOperator *Operator
```
```suggestion
	// The operator used by the search scope
	SearchOperator *Operator
```

System-Glitch commented on settings.go:
```diff
@@ -74,6 +80,28 @@ func (s *Settings) Scope(db *gorm.DB, request *goyave.Request, dest interface{})
 		}
 	}
 
+	if !s.DisableSearch && s.SearchOperator != nil && request.Has("search") {
```
Allow `s.SearchOperator` to be `nil` and use a default operator as said earlier.

System-Glitch commented on settings_test.go:
```diff
@@ -337,15 +424,15 @@ func TestApplyFilters(t *testing.T) {
 func TestSelectScope(t *testing.T) {
 	db, _ := gorm.Open(&tests.DummyDialector{}, nil)
-	db = db.Scopes(selectScope(nil, nil)).Find(nil)
+	db = db.Scopes(selectScope(nil, nil, true)).Find(nil)
 	assert.Empty(t, db.Statement.Selects)
 
 	modelIdentity := &modelIdentity{TableName: "test_models"}
 	db, _ = gorm.Open(&tests.DummyDialector{}, nil)
-	db = db.Scopes(selectScope(modelIdentity, []string{"a", "b"})).Find(nil)
+	db = db.Scopes(selectScope(modelIdentity, []string{"a", "b"}, true)).Find(nil)
 	assert.Equal(t, []string{"`test_models`.`a`", "`test_models`.`b`"}, db.Statement.Selects)
 
 	db, _ = gorm.Open(&tests.DummyDialector{}, nil)
-	db = db.Scopes(selectScope(modelIdentity, []string{})).Find(nil)
+	db = db.Scopes(selectScope(modelIdentity, []string{}, true)).Find(nil)
 	assert.Equal(t, []string{"1"}, db.Statement.Selects)
```
Add a test to check `withSelects = false` behavior.

System-Glitch commented on validation.go:
```diff
@@ -74,6 +78,14 @@ func validateJoin(ctx *validation.Context) bool {
 	return true
 }
 
+func validateSearch(ctx *validation.Context) bool {
```
Because "search" is only a string, it's not necessary. We can just use the built-in `string` rule.

Clement3 committed:
Remove Search validation, not useful

Clement3 committed:
Search - Add a default Operator in case of the developer do not define an operator

- Set to true the NewDB in the Search Session

Commit:
Check Column is a bool before apply query

Commit:
Remove the check on the column existance

Rollback to a bool to apply the Context Selects

System-Glitch committed:
Get GORM data type for column

System-Glitch committed:
Add DataType param to Operator

Commit:
Fix override in selectScope and add tests

Commit:
Add test for empty search fields

Commit:
Add a ApplySearch that return a Search struct

- Test if the Search Fields has been set by getSelectableFields
- Test if the operator is set if null by $cont operator

System-Glitch commented:
Thank you ! 🎉
